### PR TITLE
Apply transforms on 2D patches

### PR DESCRIPTION
**Pull request by mariehbourget**
#### GitHub

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've assigned a reviewer

#### PR contents

- [x] I've consulted [ivadomed's internal developer documentation](https://github.com/ivadomed/ivadomed/wiki) to ensure my contribution is in line with any relevant design decisions
- [ ] I've added [relevant tests](https://github.com/ivadomed/ivadomed/wiki/tests) for my contribution
- [ ] I've updated the [relevant documentation](https://github.com/ivadomed/ivadomed/wiki/documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description
This PR aims to apply transforms on each individual 2D patches instead of the entire image in `mri2d_segmentation_dataset.py/get_item` when loading a batch. This will partially fix performance issues described in #977 when training on large microscopy images. It also help with performance when running segmentation as described in #972.

The "slice" behavior (without patches) is unchanged.
No new test is required, all current tests are passing.
I also tested the the commands `--train`, `--test` and `--segment` in single-channel, multi-contrasts and multi-channels.

To test the PR:
The PR can be tested by training and testing a model with the addition of `length_2D` and `stride_2D` options in the config file of any `default_model` (2D unet).

## Linked issues
Related to #972 and #977.

### CODE REVIEW

File: ivadomed/loader/mri2d_segmentation_dataset.py [hunk 1]
```diff
@@ -196,81 +196,92 @@ def __getitem__(self, index):
...
+                # stack_gt is a tensor of size 1x1, values: 0 or 1
+                # "expand(1)" is necessary to be compatible with segmentation convention: n_labelxhxwxd
+                stack_gt = torch.from_numpy(seg_pair_slice["gt"][0]).expand(1)
 
-            if stack_gt is not None:
-                for _ in range(len(stack_gt)):
-                    data_dict['gt'] = stack_gt[:,
-                                       coord['x_min']:coord['x_max'],
-                                       coord['y_min']:coord['y_max']]
+        else:
```
Comment (andreanne-lemay) [inline ivadomed/loader/mri2d_segmentation_dataset.py:244]: There seems to be a lot of code duplication here between `if self.is_2d_patch` and the `else`. Could we simply keep the code under the first part of the if and simply set 
```python
if not self.is_2d_patch:
    coord = {}
    coord['x_min'], coord['x_max'], coord['y_min'], coord['y_max'] = 0, seg_pair_input["input"][0].shape[0], 0, seg_pair_input["input"][0].shape[1]
``` 
(something like that) I might have missed some technicalities why this is not possible, so please ignore if it's too complicated.

**Review by andreanne-lemay — APPROVED** [review]
Looks good overall! Makes a lot of sense.
To consider in next PRs or this one depending on how urgently we want to merge this PR:
* Do the same thing for the 3D loader.
Thanks for taking care of this!

File: ivadomed/loader/mri2d_segmentation_dataset.py (hunk 1 above)
Comment (mariehbourget) [reply ivadomed/loader/mri2d_segmentation_dataset.py:244]: I had tried to minimize the duplicated code but could not find a good way. I think your suggestion could work well, thank you!
I'll try it, the only issue I might have is to test the ROI correctly. Currently ROI only applies to entire slice and not patches and I'm not very familiar with the feature. Is the [Cascaded architecture](https://ivadomed.org/tutorials/cascaded_architecture.html#cascaded-architecture) tutorial using this feature? (by cropping around the segmented spinal cord?)

Comment (mariehbourget) [reply ivadomed/loader/mri2d_segmentation_dataset.py:201]: Thank you @andreanne-lemay!
I was able to remove the code duplication following your advice, it looks a lot better!

I answered my own question: the "cascaded" tutorial does not use the ROI feature. I ended up testing it with `ROICrop` and `suffix` in `roi_params` which I think is the right thing to do.

I tested several scenarios (Microscopy 2D with patches, MRI 2D with and without patches, and MRI with ROICrop.
Everything looks good to me and the tests are passing.

Would you be available to take a look again to make sure I did not missed something before I merge?

As for the 3D loader, I will open a separate issue as I don't have time or resources to take of it at the moment.
Thank you again!

Comment (andreanne-lemay) [reply ivadomed/loader/mri2d_segmentation_dataset.py:201]: Thanks for looking into it and sounds good for the separate 3D issue! I'm looking through it again now, sorry for the delay!

**Review by andreanne-lemay — APPROVED** [review]
LGTM! Thanks again :D

**Comment by mariehbourget** [comment]
For reference, I opened the issue #993 for the similar 3D subvolume case.

File: ivadomed/loader/mri2d_segmentation_dataset.py [hunk 2]
```diff
@@ -196,81 +196,66 @@ def __getitem__(self, index):
...
+            # ROI goes first because params of ROICrop are needed for the followings
+            stack_roi, metadata_roi = self.transform(sample=roi_pair_slice["gt"],
+                                                     metadata=metadata_roi,
+                                                     data_type="roi")
+            # Update metadata_input with metadata_roi
+            metadata_input = imed_loader_utils.update_metadata(metadata_roi, metadata_input)
+
+        # Add coordinates of slices or patches to input metadata
+        for metadata in metadata_input:
+            metadata['coord'] = [coord["x_min"], coord["x_max"],
```
Comment (cakester) [inline ivadomed/loader/mri2d_segmentation_dataset.py:217]: coord is not defined here if you didn't go into the else statement

**Pull request merged**